File: java/parsers_java21/src/main/java/jp/co/teruuu/parser/common/EscapeParser.java
package jp.co.teruuu.parser.common;

public class EscapeParser implements Parser<String> {
    @Override
    public ParseResult<String> parse(String input, int location) {
        if (input.length() >= location + 1 && input.startsWith("\\", location)) {
            char c = input.charAt(location + 1);
            if (c == ' ') {
                return new ParseResult.Success<>(String.valueOf(' '), location + 2);
            } else if (c == 't') {
                return new ParseResult.Success<>(String.valueOf('\t'), location + 2);
            } else if (c == 'f') {
                return new ParseResult.Success<>(String.valueOf('\f'), location + 2);
            } else if (c == 'b') {
                return new ParseResult.Success<>(String.valueOf('\b'), location + 2);
            } else if (c == 'r') {
                return new ParseResult.Success<>(String.valueOf('\r'), location + 2);
            } else if (c == 'n') {
                return new ParseResult.Success<>(String.valueOf('\n'), location + 2);
            } else if (c == '\\') {
                return new ParseResult.Success<>(String.valueOf('\\'), location + 2);
            } else if (c == '"') {
                return new ParseResult.Success<>(String.valueOf('"'), location + 2);
            } else if (c == '\'') {
                return new ParseResult.Success<>(String.valueOf('\''), location + 2);
            }
        }
        return new ParseResult.Failure<>(String.format("not escape location=[%d] input=[%s]", location, input), location);
    }
}

File: rust_docker/parser_back/src/core/util/escape_parser.rs
use std::rc::Rc;

use crate::core::parser::Parser;
use crate::core::parse_result::ParseResult;
use crate::core::parse_error::ParseError;

pub fn escape_parser<'a>(escape_char: char) -> Parser<'a, char, impl Fn(&'a [u8], usize) -> ParseResult<char> + Clone + 'a> {
    let cu8 = escape_char as u8;
    let f = move |input: &'a [u8], location: usize| {
        if location + 1 < input.len() && input[location] == b'\\' && input[location + 1] == cu8 {
            ParseResult::Success { value: escape_char, location: location + 2}
        } else {
            ParseResult::Failure {
                parse_error: ParseError {
                    label: "char".to_string(),
                    message: "not match".to_string(),
                    location,
                    children: vec![]
                },
                location,
            }
        }
    };
    Parser::new(Rc::new(f))
}

pub fn escape_squote<'a>() -> Parser<'a, char, impl Fn(&'a [u8], usize) -> ParseResult<char> + Clone + 'a> {
    escape_parser('\'')
}

pub fn escape_dquote<'a>() -> Parser<'a, char, impl Fn(&'a [u8], usize) -> ParseResult<char> + Clone + 'a> {
    escape_parser('"')
}

#[test]
pub fn test_escape() {

    let parser = escape_dquote();

    println!("bytes: {:?}", "ab\\\"cd".as_bytes());

    match parser.parse("ab\\\"cd".as_bytes(), 2) {
        ParseResult::Success { value, location } => {
            assert!(true);
            assert_eq!('"', value);
            assert_eq!(4, location)
        }
        _ => assert!(false),
    }

}